Support Go Modules builds.

nevkontakte opened the pull request:
Fixes https://github.com/gopherjs/gopherjs/issues/855.

I took the approach of leveraging module support that's built into the `go/build` package. Doing so has two prerequisites:

 1. `go/build.Context` instance must not have its file system methods overridden.
 2. ReleaseTags must match the default.

Most of the changes in this PR tend to the #1. We were overriding FS access methods in order to make packages embedded in the VFS available during the build. Now we use two separate contexts chained together to access packages on the real FS and the embedded VFS. This refactoring is useful in the long run to make package loading logic more encapsulated in the build package.

The second prerequisite is somewhat bogus for reasons explained in the `build/versionhack` package. https://github.com/golang/go/issues/46856 is filed in the upstream for a better behavior, but until it's fixed, we will just work around it.

aykevl commented:
Didn't look at the code, but this might be useful: https://github.com/tinygo-org/tinygo/pull/941
In TinyGo, I've basically constructed a modified GOROOT tree (in a temporary directory) and then used `go list -json -deps` directly.

nevkontakte commented:
Thanks! I actually did look at that exact PR at one point, and [materializing GOROOT](https://github.com/gopherjs/gopherjs/issues/1021) was behind the two of my first attempts to support Modules (this PR representing the third).

This did, however, prove somewhat tricky to get right:

  - Standard library augmentation GopherJS does is more complicated than TinyGo. The latter replaces whole packages (if I remember correctly), while GopherJS rewrites parts of AST. Writing this augmented AST back on disk and making sure it can be correctly parsed again revealed a whole host of problematic edge cases.
  - There are some complications around `syscall` package and how we deal with OS-dependent code introduce another layer of complexity. Fixing https://github.com/gopherjs/gopherjs/issues/693 would eliminate this problem, but at that point it looked a lot like a rabbit hole :)

You are welcome to check out https://github.com/nevkontakte/gopherjs/tree/go-tool/cmd/gopherjs-ng which has some of those experiments if you are curious.

I still like the idea of materializing augmented GOROOT on disk, and I'm going to continue working on it, but I felt like we need a working support for Modules before 1.17, which was rumored to deprecate GOPATH entirely. This is PR is the simplest solution to the problem I found, but it also doesn't address some of the issues which make our build system so complicated and weird.

nevkontakte commented:
Also worth mentioning that initially `go/build` did not support modules and was not meant to, but at some point (not sure when, but probably recently) to did get some level of it by doing that same `go list` under the hood. So maybe TinyGo implementation could be simplified a bit at this stage 🙂

flimzy commented:
This seems to fail for me when trying to build a project containing the `spf13/cobra` library.  Reproduction case in gist here: https://gist.github.com/flimzy/3485a8e31377c9b07d8622351813c0f7

nevkontakte commented:
I've fixed the panic, which was triggered when expanding patterns containing "/...". I also replaced the gotool package with an actual go tool invocation, since the package predated go modules and didn't know how to match them anyway.

aykevl commented:
>  * Standard library augmentation GopherJS does is more complicated than TinyGo. The latter replaces whole packages (if I remember correctly), while GopherJS rewrites parts of AST. Writing this augmented AST back on disk and making sure it can be correctly parsed again revealed a whole host of problematic edge cases.

Correct for TinyGo. I looked at what GopherJS did and intentionally did not do that. I made sure that only whole packages would be replaced, to simplify the build system. It makes working on the standard library for TinyGo a bit harder, but not by much. And it makes the build system a _lot_ simpler.

>   * There are some complications around `syscall` package and how we deal with OS-dependent code introduce another layer of complexity. Fixing [proposal: Limit support to one GOOS value (primarily for stdlib). #693](https://github.com/gopherjs/gopherjs/issues/693) would eliminate this problem, but at that point it looked a lot like a rabbit hole :)

That proposal seems like a very sensible thing to do!

What I do in TinyGo is that for every supported target that isn't in Go (that is, WASI and the various baremetal systems) I pick a single GOOS/GOARCH pair for the standard library to use. So for example, for WebAssembly in the browser (which is supported by the Go standard library) I use GOOS=js GOARCH=wasm as usual, but for WASI (which isn't supported by the Go standard library) I use GOOS=linux GOARCH=arm. This works very well in practice, only the syscall package and some packages that interact with the runtime (net, os, reflect, ...) need to be replaced for that to work. This means that most code in existence will think it's a different OS/arch, but in practice it'll usually run just fine. In the (few!) edge cases where it doesn't, they can use the `tinygo` build tag. One relevant PR in this context: https://github.com/tinygo-org/tinygo/pull/1964.
If this is possible in GopherJS, I can recommend this strategy. It keeps the build process much simpler and (perhaps more importantly) means that supporting a new Go version in GopherJS will be a lot simpler: TinyGo currently needs to do very little to support new Go versions and in fact supports multiple at the same time - this is at least in part due to the policy of only replacing standard library packages as a whole (so that we reduce our reliance on implementation details). For GopherJS, using GOOS=js GOARCH=wasm might be a good strategy. For users which want to differentiate between "true" js/wasm and the GopherJS version, you can simply add a `gopherjs` build tag (just like there are `gc` and `gccgo` build tags to differentiate other compilers).
That said, I do recognize that this would be a really big change which could very well break things as they are.

> Also worth mentioning that initially `go/build` did not support modules and was not meant to, but at some point (not sure when, but probably recently) to did get some level of it by doing that same `go list` under the hood. So maybe TinyGo implementation could be simplified a bit at this stage :slightly_smiling_face:

We currently call `go list` directly and this works well in practice. I don't think we'll want to change this. See https://github.com/golang/go/issues/38953 for one reason why we're using raw `go list`.

paralin commented:
On the topic of alternative package implementations, I was wondering if @aykevl  and @nevkontakte had any comment on the following idea: https://github.com/paralin/goscript

Recently there are great tools like https://esbuild.github.io/ which are native-go applications to parse and transform javascript, typescript, jsx, ... 

Go's syntax seems like it can be directly translated into TypeScript syntax, with some package replacements (as you're describing have been done in tinygo) for optimization / compat purposes

aykevl commented:
I think GoScript is getting very much off topic. I'm happy to discuss this in a different place.

nevkontakte commented:
@aykevl thanks for your detailed comment, all of what you said makes sense. In general I have high hopes that a lot of GopherJS standard library overlays could be simplified if we try to reuse wasm implementations from the upstream. That said, there's a significant challenge that is unique to GopherJS: we don't have access to the raw memory. So we have to patch any library that does pointer arithmetics, which means it's considerably more packages than `runtime`, `os`, `net` and friends. One of my biggest objectives is to make GopherJS less divergent from upstream, but that's a lot of work and I have only so many hours in a day :)

@paralin I would agree with @aykevl that this discussion would be straying quite far from this PR's subject, so how about we continue it in [#gopherjs channel](https://app.slack.com/client/T029RQSE6/C039C0R2T) in the Gopher Slack? I like the idea of goscript, but only after having worked on GopherJS compiler for a few months I began to appreciate how many subtle challenges there are if you start considering features like goto, goroutines and runtime packages.

paralin commented:
@nevkontakte Getting the following when importing syscall/js - that package is supported for compat w/ wasm code, right?

`could not import syscall/js (invalid package name: "")`

Thanks for the work on this

nevkontakte commented:
@paralin I can't reproduce the error with this minimal example: https://gist.github.com/nevkontakte/56c18ec181a10881ddd84975bd8bc782

Commands I run:

```
gopherjs build .
node run *.js
```

paralin commented:
@nevkontakte repro here: https://github.com/paralin/gopherjs-repro-1042 just run "go run ./" in the root dir.

nevkontakte commented:
@paralin I've fixed the bug, this was indeed a regression introduced by this PR. I also fixed the bug which caused CI to skip `syscall/js` package in testing.

@flimzy I think this PR is now complete. Could you take another look at it?

paralin commented:
@nevkontakte thanks for looking into it & doing the work to implement modules with gopherjs!
